library/oid.c

```diff
@@ -861,4 +895,168 @@ int mbedtls_oid_get_numeric_string(char *buf, size_t size,
     return (int) (size - n);
 }
 
+static int oid_parse_number(const char **p, const char *bound)
+{
+    int ret = MBEDTLS_ERR_ASN1_INVALID_DATA;
+    int num = 0;
+    while (*p < bound && **p >= '0' && **p <= '9') {
+        ret = 0;
+        if (num > (INT_MAX / 10)) {
+            return MBEDTLS_ERR_OID_BUF_TOO_SMALL;
+        }
+        num *= 10;
+        num += **p - '0';
+        (*p)++;
+    }
+    if (ret != 0) {
+        return ret;
+    } else {
+        return num;
+    }
+}
+
+static size_t oid_subidentifier_num_bytes(unsigned int value)
+{
+    size_t num_bytes = 1;
+    value >>= 7;
+    while (value != 0) {
+        num_bytes++;
+        value >>= 7;
+    }
+    return num_bytes;
+}
+
+static int oid_subidentifier_encode_into(unsigned char **p,
+                                         unsigned char *bound,
+                                         unsigned int value)
+{
+    size_t num_bytes = oid_subidentifier_num_bytes(value);
+    if ((size_t) (bound - *p) < num_bytes) {
+        return MBEDTLS_ERR_OID_BUF_TOO_SMALL;
+    }
+    (*p)[num_bytes - 1] = (unsigned char) (value & 0x7f);
+    value >>= 7;
+
+    for (size_t i = 2; i <= num_bytes; i++) {
+        (*p)[num_bytes - i] = 0x80 | (unsigned char) (value & 0x7f);
+        value >>= 7;
+    }
+    *p += num_bytes;
+
+    return 0;
+}
+
+/* Return the OID for the given x.y.z.... style numeric string  */
+int mbedtls_oid_from_numeric_string(mbedtls_asn1_buf *oid,
+                                    const char *buf, size_t size)
+{
+    int ret = MBEDTLS_ERR_ASN1_INVALID_DATA;
+
+    const char *str_ptr = buf;
+    const char *str_bound = buf + size;
+    int val = 0;
+    size_t encoded_len;
+
+    int component1 = oid_parse_number(&str_ptr, str_bound);
+    if (component1 < 0) {
+        return component1;
+    }
+    if (component1 > 2) {
+        /* First component can't be > 2 */
+        return MBEDTLS_ERR_ASN1_INVALID_DATA;
+    }
+    if (str_ptr >= str_bound || *str_ptr != '.') {
+        return MBEDTLS_ERR_ASN1_INVALID_DATA;
+    }
+    str_ptr++;
+
+    int component2 = oid_parse_number(&str_ptr, str_bound);
+    if (component2 < 0) {
+        return component2;
+    }
+    if ((component1 < 2) && (component2 > 38)) {
+        /* Root nodes 0 and 1 may have up to 39 children */
+        return MBEDTLS_ERR_ASN1_INVALID_DATA;
+    }
+    if (str_ptr < str_bound && *str_ptr != '\0') {
+        if (*str_ptr == '.') {
+            str_ptr++;
+        } else {
+            return MBEDTLS_ERR_ASN1_INVALID_DATA;
+        }
+    }
+
+    if ((UINT_MAX - (unsigned int) component2) <=
+        ((unsigned int) component1 * 40)) {
+        return MBEDTLS_ERR_OID_BUF_TOO_SMALL;
```

Comment:
That's not quite true, actually. `component1` must be less than or equal to 2, but if it is equal to 2 then `component2` can be as big as you like. A few OIDs use 2 bytes for the first 2 components, for example `{joint-iso-itu-t(2) cybersecurity(48)} / 2.48` or `{joint-iso-itu-t(2) uav(52)} / 2.52`.

This check is for integer overflow - a limitation in our implementation - which is denoted by the `BUF_TOO_SMALL` error code in our existing functions. I appreciate it doesn't fit very well though.

Reply:
> `component2` can be as big as you like

While that's technically true, AIUI the largest "real" component2 is `uav(52)`, and there's `example(999)` - and this code is fine with either

Reply:
But actually, `MBEDTLS_ERR_ASN1_INVALID_DATA` is a good error code for "something wrong with the data"

Reply:
And note: with `MBEDTLS_ERR_OID_BUF_TOO_SMALL` I'd expect to be able to re-size a buffer, make the call again, and have it work (i.e. a resolvable runtime error, rather than "this ain't never gonna work")

Reply:
> And note: with MBEDTLS_ERR_OID_BUF_TOO_SMALL I'd expect to be able to re-size a buffer, make the call again, and have it work (i.e. a resolvable runtime error, rather than "this ain't never gonna work")

Good point. From this perspective `MBEDTLS_ERR_ASN1_INVALID_DATA` seems more suitable for this case.

Reply:
Technically an OID subidentifier may have any value, even one that doesn't fit in 32 bits so the true error is not "invalid data", but "data that is technically valid but not handled due to the limitations of this implementation". However this seems close enough, and I haven't come across any real OID subidentifiers larger than (2^32 - 1) in any case.